assembly/src/assembler/instruction/u32_ops.rs

```diff
@@ -459,117 +348,57 @@ fn handle_division(
 fn prepare_bitwise<const MAX_VALUE: u8>(
     span: &mut SpanBuilder,
     imm: Option<u8>,
-    op_mode: U32OpMode,
-    final_ops: [Operation; 2],
 ) -> Result<Option<CodeBlock>, AssemblyError> {
-    match (imm, op_mode) {
-        (Some(0), U32OpMode::Checked) => {
-            // if shift/rotation is performed by 0, just verify that stack top is u32
-            span.push_ops([Pad, U32assert2(ZERO), Drop]);
-            return Ok(None);
-        }
-        (Some(imm), U32OpMode::Checked) => {
-            validate_param(imm, 1..=MAX_VALUE)?;
-            span.push_ops([Push(Felt::new(1 << imm)), U32assert2(ZERO)]);
-        }
-        (Some(0), U32OpMode::Unchecked) => {
+    match imm {
+        Some(0) => {
             // if shift/rotation is performed by 0, do nothing (Noop)
-            span.push_op(Noop);
-            return Ok(None);
+            span.add_op(Noop)
         }
-        (Some(imm), U32OpMode::Unchecked) => {
-            span.push_op(Push(Felt::new(1 << imm)));
+        Some(imm) => {
+            validate_param(imm, 1..=MAX_U32_ROTATE_VALUE)?;
+            span.add_op(Push(Felt::new(1 << imm)))
         }
-        (None, U32OpMode::Checked) => {
-            // Assume the dynamic shift value b is on top of the stack.
+        None => {
             append_pow2_op(span);
-            span.push_op(U32assert2(ZERO));
+            Ok(None)
         }
-        (None, U32OpMode::Unchecked) => append_pow2_op(span),
-        _ => unreachable!("unsupported operation mode"),
     }
-    span.add_ops(final_ops)
 }
 
-// COMPARISON OPERATIONS
-// ================================================================================================
-
-/// Translates u32checked_eq assembly instruction to VM operations.
-///
-/// Specifically we test the first two numbers to be u32, then perform a EQ to check the equality.
-///
-/// VM cycles per mode:
-/// - u32checked_eq: 2 cycles
-/// - u32checked_eq.b: 3 cycles
-pub fn u32eq(span: &mut SpanBuilder, imm: Option<u32>) -> Result<Option<CodeBlock>, AssemblyError> {
-    if let Some(imm) = imm {
-        push_u32_value(span, imm);
-    }
-
-    span.add_ops([U32assert2(ZERO), Eq])
-}
-
-/// Translates u32checked_neq assembly instruction to VM operations.
-///
-/// Specifically we test the first two numbers to be u32, then perform a `EQ NOT` to check the
-/// equality.
-///
-/// VM cycles per mode:
-/// - u32checked_neq: 3 cycles
-/// - u32checked_neq.b: 4 cycles
-pub fn u32neq(
+/// Adds the provided pair of final operations to the span if needed, i.e. when immediate value is
+/// not 0.
+fn add_final_ops(
     span: &mut SpanBuilder,
-    imm: Option<u32>,
+    imm: Option<u8>,
+    final_ops: [Operation; 2],
 ) -> Result<Option<CodeBlock>, AssemblyError> {
-    if let Some(imm) = imm {
-        push_u32_value(span, imm);
+    if imm != Some(0) {
+        span.add_ops(final_ops)
+    } else {
+        Ok(None)
     }
```

Comment:
I'd probably get rid of this function and just move the code into the calling functions. It will be a few extra lines of code, but I think it will be more clear.